pkg/reconcilers/external/kubeconfig.go

```diff
@@ -0,0 +1,304 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	clog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	tenancyv1alpha1 "github.com/kubestellar/kubeflex/api/v1alpha1"
+	"github.com/kubestellar/kubeflex/pkg/util"
+)
+
+const (
+	adoptedClusterSAName      = "kubeflex"
+	adoptedClusterSANamespace = "kube-system"
+)
+
+var (
+	defaultAdoptedTokenExpirationSeconds int64 = 365 * 86400
+)
+
+func (r *ExternalReconciler) ReconcileKubeconfigFromBoostrapSecret(ctx context.Context, hcp *tenancyv1alpha1.ControlPlane) error {
+	_ = clog.FromContext(ctx)
+
+	// do not reconcile if kubeconfig secret is already present
+	if r.IsKubeconfigSecretPresent(ctx, *hcp) {
+		return nil
+	}
+
+	bootstrapApiConfig, err := getKubeconfigFromBoostrapSecret(r.Client, ctx, hcp)
+	if err != nil {
+		return err
+	}
+
+	bootstrapRestConfig, err := clientcmd.NewDefaultClientConfig(*bootstrapApiConfig, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	aClientset, err := kubernetes.NewForConfig(bootstrapRestConfig)
+	if err != nil {
+		return err
+	}
+
+	if err = reconcileAdoptedClusterRoleBinding(aClientset, adoptedClusterSAName, adoptedClusterSANamespace); err != nil {
+		return fmt.Errorf("error creating ClusterRoleBinding on the adopted cluster: %v", err)
+	}
+
+	if err = reconcileAdoptedServiceAccount(aClientset.CoreV1().ServiceAccounts(adoptedClusterSANamespace), adoptedClusterSAName); err != nil {
+		return fmt.Errorf("error creating ServiceAccount on the adopted cluster: %v", err)
+	}
+
+	bearerToken, err := requestTokenWithExpiration(aClientset.CoreV1().ServiceAccounts(adoptedClusterSANamespace), adoptedClusterSAName, hcp.Spec.AdoptedTokenExpirationSeconds)
+	if err != nil {
+		return fmt.Errorf("error requesting token from the adopted cluster: %v", err)
+	}
+
+	newKubeConfig, err := createNewKubeConfig(bootstrapApiConfig, bearerToken, hcp)
+	if err != nil {
+		return fmt.Errorf("error creating adopted cluster kubeconfig: %v", err)
+	}
+
+	if err := r.ReconcileKubeconfigSecret(ctx, *hcp, newKubeConfig); err != nil {
+		return fmt.Errorf("error creating kubeconfig secret: %v", err)
+	}
+
+	return deleteBoostrapSecret(r.Client, ctx, hcp)
+}
+
+func getKubeconfigFromBoostrapSecret(crClient client.Client, ctx context.Context, hcp *tenancyv1alpha1.ControlPlane) (*api.Config, error) {
+	_ = clog.FromContext(ctx)
+
+	if hcp.Spec.BootstrapSecretRef == nil {
+		return nil, fmt.Errorf("bootstrapSecretRef must be present in the control plane")
+	}
+
+	bootstrapSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hcp.Spec.BootstrapSecretRef.Name,
+			Namespace: hcp.Spec.BootstrapSecretRef.Namespace,
+		},
+	}
+
+	err := crClient.Get(context.TODO(), client.ObjectKeyFromObject(bootstrapSecret), bootstrapSecret, &client.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	key := util.DefaultString(hcp.Spec.BootstrapSecretRef.Key, util.KubeconfigSecretKeyInCluster)
```

Comment:
Isn't the external kubeconfig the one commonly used? The whole point of adopting an external cluster is that it is external to the KubeConfig hosting cluster.